core/src/main/java/com/linecorp/armeria/common/HttpHeadersBuilder.java

```diff
@@ -0,0 +1,509 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import java.util.List;
+import java.util.Map.Entry;
+
+import javax.annotation.Nullable;
+
+/**
+ * Builds an {@link HttpHeaders}.
+ *
+ * @see HttpHeaders#builder()
+ * @see HttpHeaders#toBuilder()
+ */
+public interface HttpHeadersBuilder extends HttpHeaderGetters {
+    /**
+     * Returns a newly created {@link HttpHeaders} with the entries in this builder.
+     * Note that any further modification of this builder is prohibited after this method is invoked.
+     */
+    HttpHeaders build();
+
+    /**
+     * Specifies the hint about the number of headers which may improve the memory efficiency and performance
+     * of the underlying data structure.
+     *
+     * @return {@code this}
+     * @throws IllegalStateException if the hint was specified too late after the underlying data structure
+     *                               has been fully initialized.
+     */
+    HttpHeadersBuilder sizeHint(int sizeHint);
+
+    /**
+     * Sets whether the headers will be the last frame in an HTTP/2 stream.
+     */
+    HttpHeadersBuilder endOfStream(boolean endOfStream);
+
+    /**
+     * Sets the {@code "content-type"} header.
+     */
+    HttpHeadersBuilder contentType(MediaType contentType);
+
+    /**
+     * Removes all the headers with the specified name and returns the header value which was added first.
+     *
+     * @param name the name of the header to retrieve
+     * @return the first header value or {@code null} if there is no such header
+     */
+    @Nullable
+    String getAndRemove(CharSequence name);
+
+    /**
+     * Removes all the headers with the specified name and returns the header value which was added first.
+     *
+     * @param name the name of the header to retrieve
+     * @param defaultValue the default value
+     * @return the first header value or {@code defaultValue} if there is no such header
+     */
+    String getAndRemove(CharSequence name, String defaultValue);
+
+    /**
+     * Removes all the headers with the specified name and returns the removed header values.
+     *
+     * @param name the name of the header to retrieve
+     * @return a {@link List} of header values or an empty {@link List} if no values are found.
+     */
+    List<String> getAllAndRemove(CharSequence name);
+
+    /**
+     * Removes all the headers with the specified name and returns the header value which was added first.
+     *
+     * @param name the name of the header to search
```

Comment:
Above uses `retrieve` here users `search`. I think using `retrieve` or `get` throughout is better, `search` feels like a more heavyweight operation

Reply:
All updated to a simplified version:

```
     * @param name the header name
```